### PR TITLE
Delete Certificate for an Envoy for a Pod that was Terminated

### PR DESCRIPTION
**Pull request by draychev**
The goal of this PR is to remove certificates for Pods that have been terminated. Without this we would continue to issue certificates and keep them in cache for the life of the OSM Controller pod.

This PR introduces 2 new functions (within Mesh Catalog):
  - `releaseCertificate()`
  - `updateRelatedProxies()`

These functions will be evaluated (sequentially) when an `Announcement` of type `PodDeleted` arrives.

To help find the certificate for a given Pod (identified by its Kubernetes UID), this PR introduces `podUIDToCN` into Mesh Catalog.  This is a thread safe map, which will be populated with Pod UID to Certificate mapping.

---

This PR will fix https://github.com/openservicemesh/osm/issues/1719

After this  I am going to hand-off the work around refining and rearchitecting the OSM eventing to @eduser25  and @shashankram who will continue to hack on the signaling mechanism.

I hope that the changes I made here would be easily translated to whatever system we come up with to handle these events in the future. I deliberately did not go into future-proofing this - we could add the entire Kubernetes object in the Announcement or launch the handlers in goroutines to unblock.  I'm leaving this for another iteration (and if at all necessary).

---

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?

### CODE REVIEW

**Comment by draychev** [comment]
ref https://github.com/openservicemesh/osm/issues/1719

File: pkg/catalog/announcement_handlers.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/certificate"
+)
+
+var errEventNotHandled = errors.New("event not handled")
+
+func (mc *MeshCatalog) releaseCertificate(ann announcements.Announcement) error {
```
[the remark this answers is not in the document]
Comment (draychev) [reply pkg/catalog/announcement_handlers.go:16]: Good idea to document these! Added a docstring here!

**Review by shashankram — DISMISSED** [review]

File: pkg/catalog/proxy.go [hunk 2]
```diff
@@ -17,14 +19,20 @@ func (mc *MeshCatalog) ExpectProxy(cn certificate.CommonName) {
 }
 
 // RegisterProxy implements MeshCatalog and registers a newly connected proxy.
-func (mc *MeshCatalog) RegisterProxy(p *envoy.Proxy) {
+func (mc *MeshCatalog) RegisterProxy(proxy *envoy.Proxy) {
 	mc.connectedProxiesLock.Lock()
```
Comment (eduser25) [inline pkg/catalog/proxy.go:23]: you could very well turn this lock + map into a sync.map at this stage

Comment (draychev) [reply pkg/catalog/proxy.go:23]: Good idea - let me create another PR for that - in a few minutes!

Comment (draychev) [reply pkg/catalog/proxy.go:23]: @eduser25 Here is the PR changing the `connectedProxies` to a `sync.Map`:  https://github.com/openservicemesh/osm/pull/2061  (another 2 PRs coming)

**Review by snehachhabria — DISMISSED** [review]

**Review by eduser25 — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**